### PR TITLE
feat: Add e2e test for forkless runtime upgrade from sudo account that is multisig

### PR DESCRIPTION
**Pull request by rdoria1**

### CODE REVIEW

File: e2e-tests/js_scripts/0008-update-runtime-multisig-scheduler.js [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+    const currentRuntimeVersion = (await api.rpc.state.getRuntimeVersion()).specVersion.toNumber();
+
+    /*****************************************************************************************************
+     *************************************** SCHEDULE RUNTIME UPGRADE*************************************
+     *****************************************************************************************************/
+
+    // Retrieve the runtime to upgrade
+    const sudoAccount = await api.query.sudo.key()
+    const code = fs.readFileSync('./new-runtime/nh_runtime.wasm').toString('hex');
```
Comment (DanieleDiBenedetto) [inline e2e-tests/js_scripts/0008-update-runtime-multisig-scheduler.js:60]: Which is the version of the test runtime ? Be sure to set it high enough such that this test always pass considering we are going to normally upgrade the runtime versions

File: e2e-tests/js_scripts/0008-update-runtime-multisig-scheduler.js [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    const updateRuntimeCall = api.tx.system && api.tx.system.setCode
+        ? api.tx.system.setCode(`0x${code}`) // For newer versions of Substrate
+        : api.tx.consensus.setCode(`0x${code}`); // For previous versions
```
Comment (la10736) [inline e2e-tests/js_scripts/0008-update-runtime-multisig-scheduler.js:63]: Useless: we never supported this version

File: e2e-tests/js_scripts/0008-update-runtime-multisig-scheduler.js [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    const sudoUpdateRuntimeCall = api.tx.sudo.sudo(updateRuntimeCall)
+
+    console.log(`Upgrading from ${sudoAccount}, ${code.length / 2} bytes`);
+
+    const { block } = await api.rpc.chain.getBlock();
+    const blockNumber = block.header.number.toNumber();
+
+    const when = blockNumber + 10;
+    const priority = 100;
```
Comment (la10736) [inline e2e-tests/js_scripts/0008-update-runtime-multisig-scheduler.js:72]:
```suggestion
    const when = blockNumber + 2;
```
Maybe is better... 48 seconds faster :smile:

File: e2e-tests/0008-update-runtime-multisig-scheduler.zndsl [hunk 4]
```diff
@@ -0,0 +1,10 @@
+Description: Test of a simple transaction from Alice to Bob using a custom js script
+Network: ./network_defs/two_nodes.toml
+Creds: config
+
+alice: is up
+bob: is up
```
Comment (la10736) [inline e2e-tests/0008-update-runtime-multisig-scheduler.zndsl:6]: We can remove it: a node is enought

File: e2e-tests/js_scripts/0008-update-runtime-multisig-scheduler.js (hunk 3 above)
Comment (la10736) [reply e2e-tests/js_scripts/0008-update-runtime-multisig-scheduler.js:72]: Neremind.... sorry

File: e2e-tests/js_scripts/0008-update-runtime-multisig-scheduler.js [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    const proposal = api.tx.multisig.asMulti(
+        threshold,
+        [BOB, CHARLIE].sort(),
+        null,
+        scheduleTx.method.toHex(),
+        0
+    );
+
+    await submitExtrinsic(proposal, alice, BlockUntil.Finalized, undefined);
+
```
Comment (la10736) [inline e2e-tests/js_scripts/0008-update-runtime-multisig-scheduler.js:86]:
```suggestion
    await submitExtrinsic(proposal, alice, BlockUntil.InBlock, undefined);
```
We don't need to wait the finalized one in test

File: e2e-tests/js_scripts/0008-update-runtime-multisig-scheduler.js [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+    const approval = api.tx.multisig.asMulti(
+        threshold,
+        [ALICE, CHARLIE].sort(),
+        info.unwrapOr(null).when,
+        scheduleTx.method.toHex(),
+        paymentInfo.weight
+    );
+
+    await submitExtrinsic(approval, bob, BlockUntil.Finalized, undefined);
+
```
Comment (la10736) [inline e2e-tests/js_scripts/0008-update-runtime-multisig-scheduler.js:101]:
```suggestion
    await submitExtrinsic(approval, bob, BlockUntil.InBlock, undefined);
```

Also here

File: e2e-tests/js_scripts/0008-update-runtime-multisig-scheduler.js [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+    console.log(`multisigAddress ${Ss58MultiAddress}`);
+
+
+    /*****************************************************************************************************
+     *************************************** SET MULTISIG AS SUDO ACCOUNT ********************************
+     *****************************************************************************************************/
+
+    const newSudoKey = api.tx.sudo.setKey(Ss58MultiAddress);
+    await submitExtrinsic(newSudoKey, alice, BlockUntil.Finalized, undefined);
+
```
Comment (la10736) [inline e2e-tests/js_scripts/0008-update-runtime-multisig-scheduler.js:42]: No fork are possible ... make test faster

```suggestion
    await submitExtrinsic(newSudoKey, alice, BlockUntil.InBlock, undefined);
```

File: e2e-tests/js_scripts/0008-update-runtime-multisig-scheduler.js (hunk 3 above)
Comment (la10736) [reply e2e-tests/js_scripts/0008-update-runtime-multisig-scheduler.js:72]: Ok I did some tests use 
```suggestion
    const when = blockNumber + 5;
```

should be safe if we don't wait for finalization in all transactions

File: e2e-tests/js_scripts/0008-update-runtime-multisig-scheduler.js [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+     *****************************************************************************************************/
+
+    // Retrieve the runtime to upgrade
+    const sudoAccount = await api.query.sudo.key()
+    const code = fs.readFileSync('./new-runtime/nh_runtime.wasm').toString('hex');
+    const updateRuntimeCall = api.tx.system && api.tx.system.setCode
+        ? api.tx.system.setCode(`0x${code}`) // For newer versions of Substrate
+        : api.tx.consensus.setCode(`0x${code}`); // For previous versions
+    const sudoUpdateRuntimeCall = api.tx.sudo.sudo(updateRuntimeCall)
+
```
Comment (la10736) [inline e2e-tests/js_scripts/0008-update-runtime-multisig-scheduler.js:65]: You did it the wrong order: I think that the right one is `Multisig[Sudo[Schedule[SetCode]]]` in this order you don't need of `EnsureRootOrSudo` to update code via multisig.
`Sudo` ensure that you call `Schedule` as `Root` and `Schedule` execute scheduled action aways as `Root`.

File: e2e-tests/js_scripts/0008-update-runtime-multisig-scheduler.js (hunk 1 above)
Comment (la10736) [reply e2e-tests/js_scripts/0008-update-runtime-multisig-scheduler.js:60]: +1

I guess that generate a wasm with `spec_version = 4_000_000_000` should be something that'll work forever.

**Review by la10736 — CHANGES_REQUESTED** [review]
Really nice

- Move sudo outer of scheduler in order to don't need bout `EnsureRootOrSudo`
- Make the test faster
- Use a version that works for any future updates

**Review by DanieleDiBenedetto — APPROVED** [review]
LGTM after you apply Michele's suggestions :)

**Review by la10736 — APPROVED** [review]
LGTM

**Pull request merged**